activerecord/lib/active_record/associations/builder/belongs_to.rb

```diff
@@ -123,7 +123,17 @@ def self.define_validations(model, reflection)
       super
 
       if required
-        model.validates_presence_of reflection.name, message: :required
+        foreign_key = reflection.foreign_key.to_s
+        name = reflection.name
+
+        condition = lambda { |record|
+          record.public_send("#{name}_changed?") ||
+            record.read_attribute(foreign_key).nil? ||
+            reflection.polymorphic? ||
+            model.connection.schema_cache.foreign_keys(model.table_name).none? { |fk| fk.column.to_s == foreign_key }
```

Comment:
I think we could cache (or even eagerly compute) this last expression. 

It needs access to the schema cache, so to eagerly compute it we might need to wait for the "schema_load" event.

Reply:
Also, is this `to_s` necessary? I might be wrong by I think `ForeignKeyDefinition#column` is already a `String`.

Reply:
@casperisfine Thank you for the comments! Updated the PR with eager computation.